### PR TITLE
[SYCL] Build proper barrier deps if host task is involved in pipeline

### PR DESCRIPTION
**Pull request by KseniyaTikhomirova**

### CODE REVIEW

**Comment by KseniyaTikhomirova** [comment]
to do: 
- update windows symbols
- implement UT

File: sycl/source/handler.cpp [hunk 1]
```diff
@@ -1375,28 +1376,27 @@ void handler::use_kernel_bundle(
...
-      throw sycl::exception(
-          make_error_code(errc::invalid),
-          "Graph nodes cannot depend on events from outside the graph.");
-    }
-    if (EventGraph != Graph) {
-      throw sycl::exception(
-          make_error_code(errc::invalid),
-          "Graph nodes cannot depend on events from another graph.");
-    }
-  }
```
Comment (sergey-semenov) [inline sycl/source/handler.cpp:1394]: Could you clarify why these checks are being removed? Seems like we want to keep these diagnostics, assuming these uses are still invalid from the user's point of view.

File: sycl/unittests/scheduler/HostTaskAndBarrier.cpp [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+using ContextImplPtr = std::shared_ptr<sycl::detail::context_impl>;
+using DeviceImplPtr = std::shared_ptr<sycl::detail::device_impl>;
+
+constexpr auto DisableCleanupName = "SYCL_DISABLE_EXECUTION_GRAPH_CLEANUP";
+
+bool CheckTestExecutionRequirements(const sycl::platform &plt) {
+  if (plt.is_host()) {
+    std::cout << "Not run due to host-only environment\n";
+    return false;
+  }
```
Comment (sergey-semenov) [inline sycl/unittests/scheduler/HostTaskAndBarrier.cpp:30]: Host device is dropped so this branch won't be taken.

File: sycl/unittests/scheduler/HostTaskAndBarrier.cpp [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+  // This test only contains device image for SPIR-V capable devices.
+  if (plt.get_backend() != sycl::backend::opencl &&
+      plt.get_backend() != sycl::backend::ext_oneapi_level_zero) {
+    std::cout << "Only OpenCL and Level Zero are supported for this test\n";
+    return false;
```
Comment (sergey-semenov) [inline sycl/unittests/scheduler/HostTaskAndBarrier.cpp:35]: Won't this platform be a mock one anyway? I'm pretty sure we always treat it as if it uses one of these two backends.

**Review by sergey-semenov — COMMENTED** [review]
LGTM overall, just some minor comments.

File: sycl/include/sycl/handler.hpp [hunk 4]
```diff
@@ -3639,6 +3637,12 @@ class __SYCL_EXPORT handler {
 
   // Set that an ND Range was used during a call to parallel_for
   void setNDRangeUsed(bool Value);
+
+protected:
+  /// Registers event dependencies on this command group.
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/include/sycl/handler.hpp]: https://github.com/intel/llvm/pull/13094/commits/a9eb8a3c39e4e5d352c94f1b9b4e170f4d4c7e7b

File: sycl/source/detail/queue_impl.hpp [hunk 5]
```diff
@@ -779,8 +792,30 @@ class queue_impl {
 
       EventRet = Handler.finalize();
       EventToBuildDeps = getSyclObjImpl(EventRet);
-    } else
+    } else {
+      // The following code supports barrier synchronization if host task is
+      // involve to the scenario. Native barriers could not handle host task
+      // dependency so in case if some commands was not enqueued - blocked we
+      // track them to prevent barrier to be enqueued earlier.
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/queue_impl.hpp]: https://github.com/intel/llvm/pull/13094/commits/a9eb8a3c39e4e5d352c94f1b9b4e170f4d4c7e7b

File: sycl/source/detail/queue_impl.cpp [hunk 6]
```diff
@@ -602,6 +603,39 @@ event queue_impl::discard_or_return(const event &Event) {
   return createDiscardedEvent();
 }
 
+void queue_impl::revisitNotEnqueuedCommandsState(
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/queue_impl.cpp]: https://github.com/intel/llvm/pull/13094/commits/a9eb8a3c39e4e5d352c94f1b9b4e170f4d4c7e7b

File: sycl/source/detail/queue_impl.hpp [hunk 7]
```diff
@@ -745,14 +745,27 @@ class queue_impl {
                           std::vector<event> &MutableVec,
                           std::unique_lock<std::mutex> &QueueLock);
 
+  // Helps to manage host tasks presence in scenario with barrier usage.
+  // Approach that tracks almost all tasks to provide barrier sync for both pi
+  // tasks and host tasks is applicable for out of order queues only. Not needed
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/queue_impl.hpp]: https://github.com/intel/llvm/pull/13094/commits/a9eb8a3c39e4e5d352c94f1b9b4e170f4d4c7e7b

File: sycl/source/detail/queue_impl.hpp [hunk 8]
```diff
@@ -924,13 +959,16 @@ class queue_impl {
...
-  // Protected by common queue object mutex MMutex.
-  EventImplPtr MGraphLastEventPtr;
+  // Access should be guarded with MMutex
+  struct DependencyTrackingItems {
+    // This event is employed for enhanced dependency tracking with in-order
+    // queue
+    EventImplPtr LastEventPtr;
+    // The following two items is employed for proper out of order enqueue
+    // ordering
+    std::vector<EventImplPtr> NotEnqueuedCmdEvents;
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/queue_impl.hpp]: https://github.com/intel/llvm/pull/13094/commits/a9eb8a3c39e4e5d352c94f1b9b4e170f4d4c7e7b

File: sycl/source/detail/event_impl.hpp [hunk 9]
```diff
@@ -330,6 +335,8 @@ class event_impl {
     return MPostCompleteEvents;
   }
 
+  void updateEnqueuedState(bool IsEnqueued) { MIsEnqueued = IsEnqueued; }
```
[the remark this answers is not in the document]
Comment (KseniyaTikhomirova) [reply sycl/source/detail/event_impl.hpp]: https://github.com/intel/llvm/pull/13094/commits/a9eb8a3c39e4e5d352c94f1b9b4e170f4d4c7e7b

File: sycl/source/handler.cpp (hunk 1 above)
Comment (KseniyaTikhomirova) [reply sycl/source/handler.cpp]: should be an accident, restored https://github.com/intel/llvm/pull/13094/commits/cfb2ef092125226ae5f61e8a03e397a9fab0c298

File: sycl/unittests/scheduler/HostTaskAndBarrier.cpp (hunk 2 above)
Comment (KseniyaTikhomirova) [reply sycl/unittests/scheduler/HostTaskAndBarrier.cpp]: removed https://github.com/intel/llvm/pull/13094/commits/cfb2ef092125226ae5f61e8a03e397a9fab0c298

**Review by sergey-semenov — APPROVED** [review]

File: sycl/source/detail/queue_impl.hpp [hunk 10]
```diff
@@ -745,14 +745,27 @@ class queue_impl {
                           std::vector<event> &MutableVec,
                           std::unique_lock<std::mutex> &QueueLock);
 
+  // Helps to manage host tasks presence in scenario with barrier usage.
+  // Approach that tracks almost all tasks to provide barrier sync for both pi
+  // tasks and host tasks is applicable for out of order queues only. No-op
+  // for in order ones.
+  void tryToResetEnqueuedBarrierDep(const EventImplPtr &EnqueuedBarrierEvent);
```
Comment (bader) [inline sycl/source/detail/queue_impl.hpp:752]: This seems to be a dead code.

File: sycl/source/detail/queue_impl.hpp [hunk 11]
```diff
@@ -745,14 +745,27 @@ class queue_impl {
...
+
+  // Called on host task completion that could block some kernels from enqueue.
+  // Approach that tracks almost all tasks to provide barrier sync for both pi
+  // tasks and host tasks is applicable for out of order queues only. Not neede
```
Comment (bader) [inline sycl/source/detail/queue_impl.hpp:756]: neede -> needed

File: sycl/source/detail/queue_impl.hpp [hunk 12]
```diff
@@ -745,14 +745,27 @@ class queue_impl {
...
+  // for in order ones.
```
Comment (bader) [inline sycl/source/detail/queue_impl.hpp:757]: Considering that the backend is not aware of host tasks, how in-order property is provided for host tasks?

File: sycl/unittests/scheduler/HostTaskAndBarrier.cpp [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+  sycl::event BarrierEvent = AddTask(TestCGType::BARRIER);
+  EventImplPtr BarrierEventImpl = sycl::detail::getSyclObjImpl(BarrierEvent);
+  auto BarrierWaitList = BarrierEventImpl->getWaitList();
+  ASSERT_EQ(BarrierWaitList.size(), 0u);
+  EXPECT_EQ(BarrierEventImpl->isEnqueued(), true);
+
+  sycl::event HTEvent = AddTask(TestCGType::HOST_TASK);
+  EventImplPtr HostTaskEventImpl = sycl::detail::getSyclObjImpl(HTEvent);
+  auto HostTaskWaitList = HostTaskEventImpl->getWaitList();
+  ASSERT_EQ(HostTaskWaitList.size(), 0u);
```
Comment (bader) [inline sycl/unittests/scheduler/HostTaskAndBarrier.cpp:154]: Shouldn't host task to wait for barrier to complete?

File: sycl/unittests/scheduler/HostTaskAndBarrier.cpp [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+  EXPECT_EQ(HostTaskEventImpl->isEnqueued(), true);
+
+  sycl::event KernelEvent = AddTask(TestCGType::KERNEL_TASK);
+  EventImplPtr KernelEventImpl = sycl::detail::getSyclObjImpl(KernelEvent);
+  auto KernelWaitList = KernelEventImpl->getWaitList();
+  ASSERT_EQ(KernelWaitList.size(), 0u);
```
Comment (bader) [inline sycl/unittests/scheduler/HostTaskAndBarrier.cpp:160]: Do I understand correctly that the barrier semantics is guaranteed by the backend for the kernel task, so there is not dependency on barrier?

I suggest adding in-order queue test case with code comments describing code logic.

File: sycl/unittests/scheduler/HostTaskAndBarrier.cpp [hunk 15]
```diff
@@ -0,0 +1,272 @@
...
+  sycl::event BarrierEvent = AddTask(TestCGType::BARRIER);
+  EventImplPtr BarrierEventImpl = sycl::detail::getSyclObjImpl(BarrierEvent);
+  auto BarrierWaitList = BarrierEventImpl->getWaitList();
+  ASSERT_EQ(BarrierWaitList.size(), 0u);
+  EXPECT_EQ(BarrierEventImpl->isEnqueued(), true);
+
+  sycl::event HTEvent = AddTask(TestCGType::HOST_TASK);
+  EventImplPtr HostTaskEventImpl = sycl::detail::getSyclObjImpl(HTEvent);
+  auto HostTaskWaitList = HostTaskEventImpl->getWaitList();
+  ASSERT_EQ(HostTaskWaitList.size(), 0u);
```
Comment (bader) [inline sycl/unittests/scheduler/HostTaskAndBarrier.cpp:230]: I would expect to see a dependency on barrier event here (similar to `BarrierHandlingWithHostTask` test case).

File: sycl/unittests/scheduler/InOrderQueueSyncCheck.cpp [hunk 16]
```diff
@@ -61,12 +68,16 @@ TEST_F(SchedulerTest, InOrderQueueSyncCheck) {
...
+        MockCGH, detail::CG::CGTYPE::CodeplayHostTask, Event);
   }
   {
     LimitedHandlerSimulation MockCGH;
-    EXPECT_CALL(MockCGH, depends_on).Times(1);
-    Queue->finalizeHandler<LimitedHandlerSimulation>(MockCGH, Event);
+    EXPECT_CALL(MockCGH, depends_on(An<const sycl::detail::EventImplPtr &>()))
+        .Times(1);
+    Queue->finalizeHandler<LimitedHandlerSimulation>(
+        MockCGH, detail::CG::CGTYPE::CodeplayHostTask, Event);
```
Comment (bader) [inline sycl/unittests/scheduler/InOrderQueueSyncCheck.cpp:81]: Do we test scenarios with multiple in-order queues, where host task enqueued in one queue and event was passed as an argument to a barrier in another queue?

**Pull request merged**